// ng-rive/esm2022/lib/component/bone.mjs
import { Directive, Input, NgZone } from '@angular/core';
import { RiveCanvas } from '../canvas';
import { RiveTransformComponent } from './transform-component';
import * as i0 from "@angular/core";
import * as i1 from "../canvas";
export class RiveBone extends RiveTransformComponent {
    set length(value) {
        this.set('length', value);
    }
    constructor(zone, canvas) {
        super(zone, canvas);
    }
    getComponent(name) {
        return this.canvas.artboard?.bone(name);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.1.4", ngImport: i0, type: RiveBone, deps: [{ token: i0.NgZone }, { token: i1.RiveCanvas }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.1.4", type: RiveBone, isStandalone: true, selector: "riv-bone, [rivBone]", inputs: { length: "length" }, exportAs: ["rivBone"], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.1.4", ngImport: i0, type: RiveBone, decorators: [{
            type: Directive,
            args: [{
                    selector: 'riv-bone, [rivBone]',
                    exportAs: 'rivBone',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i0.NgZone }, { type: i1.RiveCanvas }]; }, propDecorators: { length: [{
                type: Input
            }] } });
