// ng-rive/esm2022/lib/component/node.mjs
import { Directive, Input, NgZone } from '@angular/core';
import { RiveCanvas } from '../canvas';
import { RiveTransformComponent } from './transform-component';
import * as i0 from "@angular/core";
import * as i1 from "../canvas";
export class RiveNode extends RiveTransformComponent {
    set x(value) {
        this.set('x', value);
    }
    get x() {
        return this.component?.x;
    }
    set y(value) {
        this.set('y', value);
    }
    get y() {
        return this.component?.y;
    }
    constructor(zone, canvas) {
        super(zone, canvas);
    }
    getComponent(name) {
        return this.canvas.artboard?.node(name);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.1.4", ngImport: i0, type: RiveNode, deps: [{ token: i0.NgZone }, { token: i1.RiveCanvas }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.1.4", type: RiveNode, isStandalone: true, selector: "riv-node, [rivNode]", inputs: { x: "x", y: "y" }, exportAs: ["rivNode"], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.1.4", ngImport: i0, type: RiveNode, decorators: [{
            type: Directive,
            args: [{
                    selector: 'riv-node, [rivNode]',
                    exportAs: 'rivNode',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i0.NgZone }, { type: i1.RiveCanvas }]; }, propDecorators: { x: [{
                type: Input
            }], y: [{
                type: Input
            }] } });
