// ng-rive/esm2022/lib/component/root-bone.mjs
import { Directive, Input, NgZone } from '@angular/core';
import { RiveCanvas } from '../canvas';
import { RiveTransformComponent } from './transform-component';
import * as i0 from "@angular/core";
import * as i1 from "../canvas";
export class RiveRootBone extends RiveTransformComponent {
    set x(value) {
        this.set('x', value);
    }
    set y(value) {
        this.set('y', value);
    }
    set length(value) {
        this.set('length', value);
    }
    constructor(zone, canvas) {
        super(zone, canvas);
    }
    getComponent(name) {
        return this.canvas.artboard?.rootBone(name);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.1.4", ngImport: i0, type: RiveRootBone, deps: [{ token: i0.NgZone }, { token: i1.RiveCanvas }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.1.4", type: RiveRootBone, isStandalone: true, selector: "riv-root-bone, [rivRootBone]", inputs: { x: "x", y: "y", length: "length" }, exportAs: ["rivRootBone"], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.1.4", ngImport: i0, type: RiveRootBone, decorators: [{
            type: Directive,
            args: [{
                    selector: 'riv-root-bone, [rivRootBone]',
                    exportAs: 'rivRootBone',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i0.NgZone }, { type: i1.RiveCanvas }]; }, propDecorators: { x: [{
                type: Input
            }], y: [{
                type: Input
            }], length: [{
                type: Input
            }] } });
