// ng-rive/esm2022/lib/component/transform-component.mjs
import { Directive, Input, NgZone } from '@angular/core';
import { RiveCanvas } from '../canvas';
import * as i0 from "@angular/core";
import * as i1 from "../canvas";
export class RiveTransformComponent {
    set name(name) {
        if (typeof name !== 'string')
            return;
        this.canvas.onReady().subscribe(() => {
            this.component = this.getComponent(name);
            if (!this.component)
                throw new Error(`Could not find component with name: "${name}"`);
            for (const key in this.state) {
                this.component[key] = this.state[key];
            }
        });
    }
    set scale(value) {
        this.set('scaleX', value);
        this.set('scaleY', value);
    }
    set scaleX(value) {
        this.set('scaleX', value);
    }
    set scaleY(value) {
        this.set('scaleX', value);
    }
    set rotation(value) {
        const v = typeof value === 'string' ? parseFloat(value) : value;
        if (v) {
            const rotation = Math.abs(v) > (2 * Math.PI) ? (v * (Math.PI / 180)) : v;
            this.set('rotation', rotation);
        }
    }
    constructor(zone, canvas) {
        this.zone = zone;
        this.canvas = canvas;
        this.state = {};
    }
    set(key, value) {
        this.zone.runOutsideAngular(() => {
            const v = typeof value === 'string' ? parseFloat(value) : value;
            if (typeof v === 'number') {
                if (this.component)
                    this.component[key] = v;
                else
                    this.state[key] = v;
            }
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.1.4", ngImport: i0, type: RiveTransformComponent, deps: [{ token: i0.NgZone }, { token: i1.RiveCanvas }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.1.4", type: RiveTransformComponent, inputs: { name: "name", scale: "scale", scaleX: "scaleX", scaleY: "scaleY", rotation: "rotation" }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.1.4", ngImport: i0, type: RiveTransformComponent, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: i0.NgZone }, { type: i1.RiveCanvas }]; }, propDecorators: { name: [{
                type: Input
            }], scale: [{
                type: Input
            }], scaleX: [{
                type: Input
            }], scaleY: [{
                type: Input
            }], rotation: [{
                type: Input
            }] } });
