// ng-rive/esm2022/lib/service.mjs
import RiveBuilder from '@rive-app/canvas-advanced';
import { Inject, Injectable, Optional } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { animationFrame } from './frame';
import { share } from 'rxjs/operators';
import { RIVE_FOLDER, RIVE_VERSION, RIVE_WASM } from './tokens';
import { firstValueFrom } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
export class RiveService {
    constructor(http, folder, wasmPath, version) {
        this.http = http;
        const riveVersion = version ?? '2.4.0';
        this.folder = folder ?? 'assets/rive';
        this.wasmPath = wasmPath ?? `https://unpkg.com/@rive-app/canvas-advanced@${riveVersion}/rive.wasm`;
    }
    async getRive() {
        if (!this.rive) {
            const locateFile = () => this.wasmPath;
            this.rive = await RiveBuilder({ locateFile });
            this.frame = animationFrame(this.rive).pipe(share());
        }
        return this.rive;
    }
    getAsset(asset) {
        return firstValueFrom(this.http.get(asset, { responseType: 'arraybuffer' }));
    }
    /** Load a riv file */
    async load(file) {
        // Provide the file directly
        if (typeof file !== 'string') {
            const [rive, buffer] = await Promise.all([
                this.getRive(),
                file.arrayBuffer(),
            ]);
            return rive?.load(new Uint8Array(buffer));
        }
        const asset = `${this.folder}/${file}.riv`;
        const [rive, buffer] = await Promise.all([
            this.getRive(),
            this.getAsset(asset),
        ]);
        if (!rive)
            throw new Error('Could not load rive');
        return rive.load(new Uint8Array(buffer));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.1.4", ngImport: i0, type: RiveService, deps: [{ token: i1.HttpClient }, { token: RIVE_FOLDER, optional: true }, { token: RIVE_WASM, optional: true }, { token: RIVE_VERSION, optional: true }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.1.4", ngImport: i0, type: RiveService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.1.4", ngImport: i0, type: RiveService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return [{ type: i1.HttpClient }, { type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [RIVE_FOLDER]
                }] }, { type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [RIVE_WASM]
                }] }, { type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [RIVE_VERSION]
                }] }]; } });
