// ng-rive/esm2022/lib/utils.mjs
export function toInt(value) {
    const v = typeof value === 'string' ? parseInt(value) : value;
    if (typeof v !== 'number')
        return;
    return v;
}
export function toFloat(value) {
    const v = typeof value === 'string' ? parseFloat(value) : value;
    if (typeof v !== 'number')
        return;
    return v;
}
export function toBool(value) {
    if (value === '' || value === true)
        return true;
    if (value === false)
        return false;
    return;
}
export function getAnimations(artboard) {
    const animations = [];
    if (!artboard)
        return [];
    const max = artboard.animationCount();
    for (let i = 0; i < max; i++) {
        animations.push(artboard.animationByIndex(i));
    }
    return animations;
}
export function getStateMachines(artboard) {
    const stateMachines = [];
    if (!artboard)
        return [];
    const max = artboard.stateMachineCount();
    for (let i = 0; i < max; i++) {
        stateMachines.push(artboard.stateMachineByIndex(i));
    }
    return stateMachines;
}
export function getClientCoordinates(event) {
    if (["touchstart", "touchmove"].indexOf(event.type) > -1 &&
        event.touches?.length) {
        event.preventDefault();
        return {
            clientX: event.touches[0].clientX,
            clientY: event.touches[0].clientY,
        };
    }
    else if (event.type === "touchend" &&
        event.changedTouches?.length) {
        return {
            clientX: event.changedTouches[0].clientX,
            clientY: event.changedTouches[0].clientY,
        };
    }
    else {
        return {
            clientX: event.clientX,
            clientY: event.clientY,
        };
    }
}
;
